Add JSHint to build

kdeloach opened the pull request:
This adds a new npm command to lint the project and also will fail the build if any lint is present.

Run jshint with this command:

    ./scripts/npm.sh run lint

Connects #387

lliss commented on src/mmw/js/src/core/setup.js:
```diff
@@ -7,6 +7,9 @@
 var $ = require('jquery');
 window.jQuery = window.$ = $;
 
+require('../../shim/backbone');
```
Are you sure this doesn't need to be added here so that other parts of the application have access to them. Didn't we have trouble with this before.

kdeloach replied:
The reason is subtle, but these dependencies have to be included in this order. It used to work fine, until I removed some "unused" require statements caught by jshint.

For example, `router.js` references `Backbone.Marionette` but it only requires `Backbone` (not `Marionette`). 

The problem is that `Marionette` attaches itself to `Backbone`. This means that `Backbone` always has to be required before `Marionette`. Instead of doing this in every module, I decided to explicitly include them in the correct order here.

kdeloach replied:
I'll add a comment.

lliss commented on src/mmw/js/src/router.js:
```diff
@@ -9,14 +9,19 @@ var Backbone = require('../shim/backbone'),
 var PatchedRouter = Backbone.Marionette.AppRouter.extend({
     route: function(route, name, callback) {
-        if (!_.isRegExp(route)) route = this._routeToRegExp(route);
+        if (!_.isRegExp(route)) {
```
Yay.

lliss commented:
Tests still pass so +1.

lliss commented:
I do agree with @mmcfarland, it would be nice if this showed as an instability rather than a hard failure.

lliss commented:
Also, as a general comment, it is nice to see this.

mmcfarland commented:
+1

kdeloach commented:
@azaveaci rebuild

kdeloach commented:
* Fixed the command to scan JS files recursively
  * Fixed (a lot) more lint
  * Updated Jenkins to mark build as unstable instead of failing if there is lint

caseypt commented:
The unused require flag is nice. +1

kdeloach commented:
@azaveaci rebuild
